ASCOM camera driver support

rkinnett opened the pull request:
Main changes:
1) segregated camera/mount/receiver hardware libraries
2) added 'ascom' model to camera hardware library

The ASCOM implementation is still pretty rough, but it works with several ASCOM-enabled astronomy cameras.  Tested with asi120mc, asi174mm,  qhy5Lii-m, and ASCOM camera simulator.  Needs review for style consistency and error handling.

rkinnett committed:
Added ASCOM driver support

rkinnett committed:
added ASCOM alt/az mount support

rkinnett committed:
Revert "added ASCOM alt/az mount support"

This reverts commit cb23ccfb1228985bea9be627b932be7192b4fb96.

gustavmpettersson committed:
refactor hw

gustavmpettersson committed:
Update hw docs

gustavmpettersson commented on pypogs/hardware_cameras.py:
```diff
@@ -248,13 +266,18 @@ def identity(self):
         - For model *ptgrey* this is the serial number *as a string*
         - Must set before initialising the device and may not be changed for an initialised device.
+        
+        - For model *ascom*, a driver name may be specified if known, (i.e. DSLR, ASICamera1, ASICamera2, 
+        QHYCCD, QHYCCD_GUIDER, QHYCCD_CAM2, AtikCameras, AtikCameras2, etc), otherwise the ASCOM driver
+        selector will open.
```
Does ascom have any way of identifying individual hardware resources? E.g. what happens if I have two cameras connected?

gustavmpettersson replied:
Does ascom have any way of identifying individual hardware resources? E.g. what happens if I have two cameras connected?

gustavmpettersson commented on pypogs/hardware_cameras.py:
```diff
@@ -139,9 +141,11 @@ def __init__(self, model=None, identity=None, name=None, auto_init=True, debug_f
         if name is not None:
             self.name = name
-        if auto_init and not None in (model, identity):
+        if auto_init and model is not None:
```
Check if model is ptgrey, if true must check identity is not None before auto init.

gustavmpettersson reviewed (changes requested):
See some specific code comments.

Remove all use of hasattr, instead set all self._ascom variables necessary to None in the constructor.

Several try-except blocks are catch-all, must be fixed.

For properties that can not be set, raise AttributeError instead of returning False.

Finally the big-ticket item, there is here a thread constantly triggering and reading out the ascom camera. Is there no image event support in ascom? The present implementation is going to be very slow.

Please work on these and make a new pull request onto hw_ascom branch.

rkinnett committed:
added TLE lookup

rkinnett committed:
trying to get rate control working

rkinnett committed:
added rate inversion control

rkinnett committed:
fixing rate control for ioptron mount

rkinnett committed:
overhauled

Changed device model selectin to ComboBox pull-down;  fixed ascom camera gain/exposure setting methods;  various fixed tested with iOptron AZMP and asi174 cameras

rkinnett committed:
Merge branch 'sandbox'

rkinnett committed:
fixed camera gain setting method

rkinnett committed:
various improvements

rkinnett committed:
Added sidereal tracking, fixed binning

rkinnett committed:
minor edits

rkinnett committed:
try to fix com threading issue and cleanup device identity handling

rkinnett committed:
updated my own startup script

rkinnett committed:
Added iOptron protocol and fixed CoInitialize error

rkinnett committed:
fixed minor error

rkinnett committed:
incorporating ephem lookup

rkinnett committed:
integrated ephem lookup

rkinnett committed:
improved device init arg parsing and improved stop behaviors

rkinnett committed:
revised start template with improved device parameter handling

rkinnett committed:
cleanup and robustness improvements

rkinnett committed:
Merge branch 'sandbox'

rkinnett committed:
minor improvements and bug fixes

rkinnett committed:
Robustness improvements with AZMP

rkinnett committed:
updated requirements.txt and other minor mods

rkinnett committed:
fix satellitetle import

Commit:
fix pywin32 import

rkinnett committed:
expose smoothing parameter in Tracker settings

rkinnett committed:
Merge branch 'master' of https://github.com/rkinnett/pypogs

rkinnett committed:
add tracking settings to example startup script and improve control settings GUI warnings

rkinnett committed:
removed ascom 8bit casting

Should work with 8 or 16 bit cameras now.

rkinnett committed:
added lines to enable recording coarse images

rkinnett committed:
return unhandled message for ascom camera frame rate setter

rkinnett committed:
improvements to auto alignment

rkinnett committed:
auto align get new image

rkinnett committed:
improvements to self alignment and loading alignment

rkinnett committed:
merged tetra3_update from esa

rkinnett committed:
exclude star catalogs and custom tetra3 databases

rkinnett committed:
don't stop star alignment if can't solve one position

rkinnett committed:
fix auto alignment settle time and failure

rkinnett committed:
minor details

rkinnett committed:
fixed serial hang problem

gustavmpettersson committed:
Merge branch 'master' into hw_ascom

rkinnett committed:
Update submodule link

rkinnett committed:
pass more properties to star camera

rkinnett committed:
Update tetra3

rkinnett committed:
added star alignment failure tolerance parameter and improved tetra3 invocation in star alignment

rkinnett committed:
Project target intercept position for initial tracking slew

rkinnett committed:
fix target projection at start of tracking

rkinnett committed:
fixed initial slew at start of tracking

rkinnett committed:
minor fixes for mount robustness

rkinnett committed:
add telescope server for stellarium

rkinnett committed:
fixing issues per 3/29/22 discussion

rkinnett committed:
fix ascom mount deinit

gustavmpettersson committed:
Initial zwoasi support (work in progress)

gustavmpettersson committed:
Add zwoasi license notice

gustavmpettersson committed:
Implemented zwoasi property setting

rkinnett committed:
add add-offset to star camera

rkinnett committed:
misc cleanup

rkinnett committed:
Merge remote-tracking branch 'upstream/hw_ascom'

rkinnett committed:
Revert "Merge remote-tracking branch 'upstream/hw_ascom'"

This reverts commit 70ed23b5604f7a07a6d205f706b8cc2fa9dcf1ac, reversing
changes made to 5fd965c1345e9a5c31e5af369eef286fc5812ec6.

rkinnett committed:
hardware subdir

rkinnett committed:
hardware subdir

rkinnett committed:
hardware subdir

rkinnett committed:
ignore tetra3

rkinnett committed:
integrated zwoasi from esa/pypogs/hw_ascom

rkinnett committed:
Merge branch 'master' of https://github.com/rkinnett/pypogs

rkinnett committed:
ignore hardware log

rkinnett committed:
improve zwoasi interfaces

rkinnett committed:
ignore log files

rkinnett committed:
fix hardware popup clear on removal

rkinnett committed:
pull in bayer handling from hw_ascom

rkinnett committed:
make zwoasi default model

rkinnett committed:
fix hardware log file dir and misc other adjustments

rkinnett committed:
change example gui update rate

rkinnett committed:
add exposure control and generally resynch with hw_ascom

rkinnett committed:
fix plate scale auto calc

rkinnett committed:
fix readout size and misc other cleanup

rkinnett committed:
fixing minor bugs

rkinnett committed:
add common sats selection

rkinnett committed:
added saved targets selection

rkinnett committed:
try to improve exposure setting robustness

rkinnett committed:
suppress img prints in gui log

rkinnett committed:
suppress extraneous logging

rkinnett committed:
add auto-align max trials parameter

rkinnett committed:
rearrange target popup

rkinnett committed:
add target server

enables TCP link from SkyTrack

rkinnett committed:
update server start commands

rkinnett committed:
Clarify differences between this fork and main project

rkinnett committed:
slight tweaks to readme

rkinnett committed:
add zwoasi

rkinnett committed:
minor tweaks

rkinnett committed:
reorganize templates

rkinnett committed:
resynch with parent project hw_ascom

rkinnett committed:
fix azmp serial interfaces

rkinnett committed:
minor bug fixes

rkinnett committed:
fix azmp normal mode state lookup

rkinnett committed:
add tracking settings adjustments

rkinnett committed:
Update mount_only.py

rkinnett committed:
minor cleanup

rkinnett committed:
improve serial port init robustness and error reporting

rkinnett committed:
add rate avoidance

rkinnett committed:
fix rate avoidance

rkinnett committed:
add projection time delta feedback setting

rkinnett committed:
reenable ASCOM chooser

rkinnett committed:
Update README.rst
